File: bcs-ui/backend/kube_core/toolkit/kubectl/kubeconfig.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

Module for generating kubeconfig file
"""
import contextlib
import logging
import tempfile
from dataclasses import dataclass
from typing import List

import yaml

logger = logging.getLogger(__name__)


@dataclass
class Cluster:
    """Cluster Info

    :param cert: path of certification
    :param cert_data: certification in base64 format, will be ignored if cert is provided
    """

    name: str
    server: str
    cert: str = ""
    cert_data: str = ""
    api_version: str = "v1"


@dataclass
class User:
    """Kubernetes user info"""

    name: str
    token: str


@dataclass
class Context:
    """kube config context"""

    name: str
    user: User
    cluster: Cluster


class KubeConfig:
    """The kubeconfig class"""

    def __init__(self, contexts: List[Context]):
        assert contexts, "Must provide at least one context"
        self.contexts = contexts

    @staticmethod
    def format_cluster(cluster):
        """Format a cluster as kubeconfig format"""
        cert_info = {"certificate-authority-data": cluster.cert_data}
        if cluster.cert:
            cert_info = {"certificate-authority": cluster.cert}

        cert_info = {"insecure-skip-tls-verify": True}

        return {
            "name": cluster.name,
            "cluster": {"server": cluster.server, "api-version": cluster.api_version, **cert_info},
        }

    @staticmethod
    def format_user(user):
        """Format an user as kubeconfig format"""
        return {"name": user.name, "user": {"token": user.token}}

    @staticmethod
    def format_context(context):
        """Format a context as kubeconfig format"""
        return {
            "name": context.name,
            "context": {
                "user": context.user.name,
                "cluster": context.cluster.name,
            },
        }

    def dumps(self, current_context: str = ""):
        """Represent current config as a kubeconfig file

        :returns: kubeconfig file content
        """
        clusters = {}
        users = {}
        for context in self.contexts:
            clusters[context.cluster.name] = self.format_cluster(context.cluster)
            users[context.user.name] = self.format_user(context.user)

        current_context = current_context or self.contexts[0].name
        payload = {
            "apiVersion": "v1",
            "kind": "Config",
            "clusters": list(clusters.values()),
            "users": list(users.values()),
            "contexts": [self.format_context(context) for context in self.contexts],
            "current-context": current_context,
        }
        return yaml.dump(payload, default_flow_style=False)

    @contextlib.contextmanager
    def as_tempfile(self):
        """A context manager which dump current config to a temp kubeconfig file"""

        with tempfile.NamedTemporaryFile() as fp:
            fp.write(self.dumps().encode())
            fp.flush()
            yield fp.name

File: bcs-services/bcs-federation-manager/internal/kubeconfig/kubeconfig.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package kubeconfig ...
package kubeconfig

import (
	"strings"

	"gopkg.in/yaml.v2"
)

const (
	// DefaultKubeconfigUserToken default user token which is useless for now,
	// but cannot be empty for register to cluster manager
	DefaultKubeconfigUserToken = "xxxxxxx"
)

// NewConfigForRegister create a new kubeconfig for register federation cluster
func NewConfigForRegister(address string) *Config {
	return newConfig(true, address, DefaultKubeconfigUserToken, "fed-cluster", "fed-context", "admin")
}

// NewConfigForProvider create a new kubeconfig for clusternet provider
func NewConfigForProvider(address string, token string, clusterId string) *Config {
	clusterId = strings.ToLower(clusterId)
	return newConfig(false, address, token, clusterId, "master-"+clusterId, "clusternet-computing")
}

func newConfig(skipTLS bool, address, token, clusterName, contextName, userName string) *Config {
	config := &Config{
		APIVersion:     "v1",
		Clusters:       make([]Cluster, 0),
		Contexts:       make([]Context, 0),
		CurrentContext: "",
		Kind:           "Config",
		Preferences:    struct{}{},
		Users:          make([]User, 0),
	}

	config.AddCluster(
		&Cluster{
			Name: clusterName,
			Cluster: ClusterInfo{
				InsecureSkipTLSVerify: skipTLS,
				Server:                address,
			},
		},
	).AddContext(
		&Context{
			Name: contextName,
			Context: ContextInfo{
				Cluster: clusterName,
				User:    userName,
			},
		},
	).AddUser(
		&User{
			Name: userName,
			User: UserInfo{
				Token: token,
			},
		},
	).SetCurrentContext(contextName)

	return config
}

// Config kubeconfig for import cluster
type Config struct {
	APIVersion     string    `yaml:"apiVersion"`
	Clusters       []Cluster `yaml:"clusters"`
	Contexts       []Context `yaml:"contexts"`
	CurrentContext string    `yaml:"current-context"`
	Kind           string    `yaml:"kind"`
	Preferences    struct{}  `yaml:"preferences"`
	Users          []User    `yaml:"users"`
}

// Yaml return the yaml format string
func (c *Config) Yaml() string {
	result, _ := yaml.Marshal(c)
	return string(result)
}

// AddCluster add a cluster to kubeconfig
func (c *Config) AddCluster(cls *Cluster) *Config {
	if c.Clusters == nil {
		c.Clusters = make([]Cluster, 0)
	}
	c.Clusters = append(c.Clusters, *cls)
	return c
}

// AddContext add a context to kubeconfig
func (c *Config) AddContext(ctx *Context) *Config {
	if c.Contexts == nil {
		c.Contexts = make([]Context, 0)
	}
	c.Contexts = append(c.Contexts, *ctx)
	return c
}

// AddUser add a user to kubeconfig
func (c *Config) AddUser(usr *User) *Config {
	if c.Users == nil {
		c.Users = make([]User, 0)
	}
	c.Users = append(c.Users, *usr)
	return c
}

// SetCurrentContext set the current context
func (c *Config) SetCurrentContext(context string) *Config {
	c.CurrentContext = context
	return c
}

// Cluster represents a cluster configuration
type Cluster struct {
	Cluster ClusterInfo `yaml:"cluster"`
	Name    string      `yaml:"name"`
}

// ClusterInfo represents the details of a cluster
type ClusterInfo struct {
	InsecureSkipTLSVerify bool   `yaml:"insecure-skip-tls-verify,omitempty"`
	Server                string `yaml:"server"`
}

// Context represents a context configuration
type Context struct {
	Context ContextInfo `yaml:"context"`
	Name    string      `yaml:"name"`
}

// ContextInfo represents the details of a context
type ContextInfo struct {
	Cluster string `yaml:"cluster"`
	User    string `yaml:"user"`
}

// User represents a user configuration
type User struct {
	Name string   `yaml:"name"`
	User UserInfo `yaml:"user"`
}

// UserInfo represents the details of a user
type UserInfo struct {
	Token string `yaml:"token"`
}
